src/Canvas.tsx

```diff
@@ -0,0 +1,102 @@
+import React from 'react'
+import { Point } from './Point'
+import { Viewport } from './Viewport';
+import { emitKeypressEvents } from 'readline';
+
+interface ICanvasProps {
+  viewport: Viewport
+  updateViewport: (viewport: Viewport) => any,
+}
+
+interface ICanvasState {
+
+}
+
+export class Canvas extends React.Component <ICanvasProps, ICanvasState> {
+  svgRef: React.RefObject<SVGSVGElement>
+  gRef: React.RefObject<SVGGElement>
+  panActive: boolean = false
+
+  constructor(props: Readonly<ICanvasProps>) {
+    super(props)
+    this.onPanStart = this.onPanStart.bind(this)
+    this.onPanMove = this.onPanMove.bind(this)
+    this.onPanEnd = this.onPanEnd.bind(this)
+    this.onZoom = this.onZoom.bind(this)
+    this.svgRef = React.createRef();
+    this.gRef = React.createRef();
+  }
+
+  private onPanMove(e: React.MouseEvent<SVGGElement, MouseEvent>) {
+    const { viewport } = this.props
+    if (this.panActive && e.currentTarget) {
+      viewport.center = new Point(
+        viewport.center.x + ((e.movementX / viewport.gridUnit)),
+        viewport.center.y + ((e.movementY / viewport.gridUnit))
+      )
+      this.props.updateViewport(viewport);
+    }
+  }
+
+  onPanEnd() {
+    this.panActive = false;
+  }
+
+  onPanStart() {
+    this.panActive = true
+  }
+
+  private onZoom(event: React.MouseEvent<SVGGElement, MouseEvent>) {
+    const { viewport } = this.props;
+    // TODO: according to docs `event.deltaY` is not stable, but it works fine so i've tested so far, consider using scroll events if anybody will experience improper behaviour
+    // https://developer.mozilla.org/en-US/docs/Web/API/Element/wheel_event
+    const factor = 1 / viewport.gridUnit
+    // @ts-ignore
+    const zoomChange = (event.deltaY < 0 ? 1 : -1) * factor;
+    const zoomLevel = Math.max(factor, viewport.zoomLevel + zoomChange);
+    if (zoomLevel !== viewport.zoomLevel && this.svgRef.current !== null) {
+      const clientRect = this.svgRef.current.getBoundingClientRect();
+      const center = new Point(clientRect.width / 2, clientRect.height / 2)
+      const scaleChange = (1 / zoomLevel) - (1 / viewport.zoomLevel)
+      const offset = new Point(
+        ((event.clientX - center.x) * scaleChange * -1),
+        ((event.clientY - center.y) * scaleChange * -1),
+      )
+      const gridOffset = viewport.pixelToGrid(offset)
+      viewport.center = viewport.center.addition(gridOffset);
```

Comment:
zoom to point functionality doesn't work

Reply:
Yes, I got some of the math wrong somewhere -- it's moving in the right direction but not to the exact place. My goal was to make the `viewBox` values be logical and to move this to run within the SVG context only, because it has a different coordinate system. But there's a multiplier missing somewhere.